src/Program.cs

```diff
@@ -0,0 +1,471 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Text;
+using McMaster.Extensions.CommandLineUtils;
+using Sprache;
+using Newtonsoft.Json;
+
+
+namespace secondtry
+{
+    class Program
+    {
+        static void Main(string[] args)
+        {
+            commands obj = new commands();
+
+            obj.Idel(args);
+        }
+    }
+
+    class commands
+    {
+        public int Idel(string[] commands)
+        {
+            var app = new CommandLineApplication();
+            var helptemplate = "-h|--help";
+            app.HelpOption(helptemplate);
+            app.Command("use", u =>
+            {
+                u.HelpOption(helptemplate);
+                u.Description = "Dieser Befehl soll es ermöglichen die hinterlegte Konfiguration zu editieren.";
+                var path = u.Option(@"--path <fullpath>", "Setzt einen benutzerdefinierten Pfad, wenn dieser Befehl nicht benutzt wird, wird der Sampelsordner benutzt.", CommandOptionType.SingleValue);
+                u.OnExecute(() => { useon(path); });
+            });
+            return app.Execute(commands);
+        }
+
+        public void setuserpath(string mypath)
+        {
+            string[] userconfig = File.ReadAllLines(@"..\\..\\..\\..\\..\\config\\qwertz.json");
+            StringBuilder newFile = new StringBuilder();
+            string[] file = File.ReadAllLines($@"..\\..\\..\\..\\..\\config\\qwertz.json");
+            List<string> list = new List<string>(file);
+            list[2] = "\"userpath\": " + "\"" + mypath + "\"";
+            foreach (string line in list)
+            {
+                newFile.Append(line + "\n");
+            }
+            File.WriteAllText($@"..\\..\\..\\..\\config\\qwertz.json", newFile.ToString());
+        }
+        public void useon(CommandOption path) // neue funktion -> suchen und ersetzten 
```

Comment:
Ab hier wird es gruslig. 

Hier gibt es gleich mehrere Punkte die nicht so gelungen sind.

Zuerst der Name dieser Funktion. Gib ihr doch ein sprechenden Namen.  z.b. `OnExecute` oder `run` oder sowas.

desweiteren beginnt hier das verstecken der eigentlichen business Logik. Aber hier ist nicht mehr nach vollziehbar was passiert weil du hier Funktionen verschachtelt hast.

Das meint du hast das gemacht:

useon ruft openfiles auf dieses ruft parseinobject auf und dieses macht irgendwann ein replace files.

Aus leser Sicht ist das aber echt kacke, weil man in der useon funktion erstmal nur denk das Dateien geöffnet werden. Evtl. noch eingelese. Aber von mehr geht man nicht aus.

dies meint baue das ganze lieber so:

```csharp
var filestream = opnefiles(path);
var configs = parseInput(filesstream);
configs = replaceContent(configs);
replaceFiels(configs, outputpath);
```

Somit ist sichergestellt das es immer nachvollziehbar ist was passiert.

Reply:
Anschließend an dieses Feedback solltest du unbedingt auch die complexität der einzelnen Funktionen reduzieren indem du dieses Aufsplittest in kleinere. Eine gutes Indiz für schlecht code ist meistens wie tief er verschachtelt ist. Und gerade in der Methode `parseinobject` ist es sehr extrem.

Dort hast du teilweise bis zu 9 Verschachtlungen von foreach, while, switch case, if etc...

Das macht code in der regel sehr schwer lesbar.

In der Regel kann man diese vereinfachen in dem man einfach den Content der Verschachtelung in eine Funktion auslagert. Hier mal ein Beispiel:

Du hast diese If in der besagten Funktion.

```csharp
 if (subidentexit){
   if (ParserGrammarIn.getsubident.Parse(line) == "network-dev" ||
      ParserGrammarIn.getsubident.Parse(line) == "interface network-if" ||
      ParserGrammarIn.getsubident.Parse(line) == "proxy-set" ||
      ParserGrammarIn.getsubident.Parse(line) == "proxy-ip" ||
      ParserGrammarIn.getsubident.Parse(line) == "exit")
   {
      if (ParserGrammarIn.getsubident.Parse(line) == "exit"){
         configureexit = true;
         continue;
      }
      subidentexit = false;
      subident = ParserGrammarIn.getsubident.Parse(line);
      subidentvalue = ParserGrammarIn.subidentvalue.Parse(line);
      continue;
   }
}
else
{
   ...
}

```

Dies könnte man sehr viel einfacher ausdrücken wenn man hier einfach eine treffende Funktion schreibt die für sich geschrieben funktioniert:

```csharp

private void getIdentNameAndValue(string line, bool configureExit, bool subIdentExit, out string subIdentName, out string subIdentValue)
        {
            subIdentName = ParserGrammarIn.getsubident.Parse(line);
            subIdentValue = String.Empty;
            if (subIdentName == "network-dev" ||
                subIdentName == "interface network-if" ||
                subIdentName == "proxy-set" ||
                subIdentName == "proxy-ip" ||
                subIdentName == "exit")
            {
                if (subIdentName == "exit")
                {
                    configureExit = true;
                    return;
                }
                subIdentExit = false;
                subIdentValue = ParserGrammarIn.subidentvalue.Parse(line);
            }
        }
```

Reply:
du hast außerdem sehr oft diese Zeilen verwendet:

```csharp
foreach (var propertyInfo in item.GetType().GetProperties())
{
   var value = propertyInfo.GetValue(item);
   if (value != null)
   {
      i.GetType().GetProperty(propertyInfo.Name).SetValue(i, value);
   }
}
```

Verallgemeinern diesen Fall doch einfach mit einer Funktion. So kannst dir viel Code sparen.

Reply:
wird gemacht :D

Reply:
gerade gesehen das ich noch ein stück mehr verallgemeinern kann ... das mach ich morgen :)